File: tmp_cfr/decompiled/org/cobalt/api/util/player/MovementManager.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  kotlin.Metadata
 *  kotlin.jvm.JvmField
 *  kotlin.jvm.JvmStatic
 *  org.jetbrains.annotations.NotNull
 */
package org.cobalt.api.util.player;

import kotlin.Metadata;
import kotlin.jvm.JvmField;
import kotlin.jvm.JvmStatic;
import org.jetbrains.annotations.NotNull;

@Metadata(mv={2, 3, 0}, k=1, xi=48, d1={"\u0000\u001a\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0010\u000b\n\u0000\n\u0002\u0010\u0002\n\u0002\b\u001c\b\u00c6\u0002\u0018\u00002\u00020\u0001B\t\b\u0002\u00a2\u0006\u0004\b\u0002\u0010\u0003J\u0019\u0010\u0007\u001a\u00020\u00062\b\b\u0002\u0010\u0005\u001a\u00020\u0004H\u0007\u00a2\u0006\u0004\b\u0007\u0010\bJ\u0019\u0010\t\u001a\u00020\u00062\b\b\u0002\u0010\u0005\u001a\u00020\u0004H\u0007\u00a2\u0006\u0004\b\t\u0010\bJG\u0010\u0011\u001a\u00020\u00062\u0006\u0010\n\u001a\u00020\u00042\u0006\u0010\u000b\u001a\u00020\u00042\u0006\u0010\f\u001a\u00020\u00042\u0006\u0010\r\u001a\u00020\u00042\u0006\u0010\u000e\u001a\u00020\u00042\u0006\u0010\u000f\u001a\u00020\u00042\u0006\u0010\u0010\u001a\u00020\u0004H\u0007\u00a2\u0006\u0004\b\u0011\u0010\u0012J\u000f\u0010\u0013\u001a\u00020\u0006H\u0007\u00a2\u0006\u0004\b\u0013\u0010\u0003R\u0016\u0010\u0014\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u0014\u0010\u0015R\u0016\u0010\u0016\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u0016\u0010\u0015R\u0016\u0010\u0017\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u0017\u0010\u0015R\u0016\u0010\u0018\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u0018\u0010\u0015R\u0016\u0010\u0019\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u0019\u0010\u0015R\u0016\u0010\u001a\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u001a\u0010\u0015R\u0016\u0010\u001b\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u001b\u0010\u0015R\u0016\u0010\u001c\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u001c\u0010\u0015R\u0016\u0010\u001d\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u001d\u0010\u0015R\u0016\u0010\u001e\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u001e\u0010\u0015R\u0016\u0010\u001f\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b\u001f\u0010\u0015R\u0016\u0010 \u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b \u0010\u0015R\u0016\u0010!\u001a\u00020\u00048\u0006@\u0006X\u0087\u000e\u00a2\u0006\u0006\n\u0004\b!\u0010\u0015\u00a8\u0006\""}, d2={"Lorg/cobalt/api/util/player/MovementManager;", "", "<init>", "()V", "", "state", "", "setLookLock", "(Z)V", "setMovementLock", "forward", "backward", "left", "right", "jump", "shift", "sprint", "setForcedMovement", "(ZZZZZZZ)V", "clearForcedMovement", "isLookLocked", "Z", "isMovementLocked", "hasForcedMovement", "forcedForward", "forcedBackward", "forcedLeft", "forcedRight", "forcedJump", "forcedShift", "forcedSprint", "forcedAttack", "forcedUse", "forcedActionsEnabled", "cobalt"})
public final class MovementManager {
    @NotNull
    public static final MovementManager INSTANCE = new MovementManager();
    @JvmField
    public static volatile boolean isLookLocked;
    @JvmField
    public static volatile boolean isMovementLocked;
    @JvmField
    public static volatile boolean hasForcedMovement;
    @JvmField
    public static volatile boolean forcedForward;
    @JvmField
    public static volatile boolean forcedBackward;
    @JvmField
    public static volatile boolean forcedLeft;
    @JvmField
    public static volatile boolean forcedRight;
    @JvmField
    public static volatile boolean forcedJump;
    @JvmField
    public static volatile boolean forcedShift;
    @JvmField
    public static volatile boolean forcedSprint;
    @JvmField
    public static volatile boolean forcedAttack;
    @JvmField
    public static volatile boolean forcedUse;
    @JvmField
    public static volatile boolean forcedActionsEnabled;

    private MovementManager() {
    }

    @JvmStatic
    public static final void setLookLock(boolean state) {
        isLookLocked = state;
    }

    public static /* synthetic */ void setLookLock$default(boolean bl, int n, Object object) {
        if ((n & 1) != 0) {
            bl = true;
        }
        MovementManager.setLookLock(bl);
    }

    @JvmStatic
    public static final void setMovementLock(boolean state) {
        isMovementLocked = state;
        if (!state) {
            MovementManager.clearForcedMovement();
        }
    }

    public static /* synthetic */ void setMovementLock$default(boolean bl, int n, Object object) {
        if ((n & 1) != 0) {
            bl = true;
        }
        MovementManager.setMovementLock(bl);
    }

    @JvmStatic
    public static final void setForcedMovement(boolean forward, boolean backward, boolean left, boolean right, boolean jump, boolean shift, boolean sprint) {
        forcedForward = forward;
        forcedBackward = backward;
        forcedLeft = left;
        forcedRight = right;
        forcedJump = jump;
        forcedShift = shift;
        forcedSprint = sprint;
        hasForcedMovement = true;
    }

    @JvmStatic
    public static final void clearForcedMovement() {
        hasForcedMovement = false;
        forcedForward = false;
        forcedBackward = false;
        forcedLeft = false;
        forcedRight = false;
        forcedJump = false;
        forcedShift = false;
        forcedSprint = false;
        forcedAttack = false;
        forcedUse = false;
        forcedActionsEnabled = false;
    }
}

File: src/main/kotlin/org/phantom/api/util/player/MovementManager.kt
package org.phantom.api.util.player

object MovementManager {

  @JvmField
  @Volatile
  var isLookLocked = false

  @JvmField
  @Volatile
  var isMovementLocked = false

  @JvmField
  @Volatile
  var hasForcedMovement = false

  @JvmField
  @Volatile
  var forcedForward = false

  @JvmField
  @Volatile
  var forcedBackward = false

  @JvmField
  @Volatile
  var forcedLeft = false

  @JvmField
  @Volatile
  var forcedRight = false

  @JvmField
  @Volatile
  var forcedJump = false

  @JvmField
  @Volatile
  var forcedShift = false

  @JvmField
  @Volatile
  var forcedSprint = false

  @JvmField
  @Volatile
  var forcedAttack = false

  @JvmField
  @Volatile
  var forcedUse = false

  @JvmField
  @Volatile
  var forcedActionsEnabled = false

  @Volatile
  var movementOwner: MovementOwner = MovementOwner.NONE
    private set

  @Volatile
  var lookOwner: MovementOwner = MovementOwner.NONE
    private set

  @JvmStatic
  fun canTakeMovement(owner: MovementOwner): Boolean {
    val current = movementOwner
    return current == MovementOwner.NONE ||
      current == owner ||
      owner.priority >= current.priority
  }

  @JvmStatic
  fun canTakeLook(owner: MovementOwner): Boolean {
    val current = lookOwner
    return current == MovementOwner.NONE ||
      current == owner ||
      owner.priority >= current.priority
  }

  @JvmStatic
  fun requestMovement(owner: MovementOwner): Boolean {
    if (!canTakeMovement(owner)) return false
    movementOwner = owner
    isMovementLocked = true
    return true
  }

  @JvmStatic
  fun requestLook(owner: MovementOwner): Boolean {
    if (!canTakeLook(owner)) return false
    lookOwner = owner
    isLookLocked = true
    return true
  }

  @JvmStatic
  fun releaseMovement(owner: MovementOwner) {
    if (owner != MovementOwner.PATHFINDER && owner != MovementOwner.FAILSAFE && movementOwner != owner) return
    movementOwner = MovementOwner.NONE
    isMovementLocked = false
    clearForcedMovement()
  }

  @JvmStatic
  fun releaseLook(owner: MovementOwner) {
    if (owner != MovementOwner.PATHFINDER && owner != MovementOwner.FAILSAFE && lookOwner != owner) return
    lookOwner = MovementOwner.NONE
    isLookLocked = false
  }

  @JvmStatic
  fun setLookLock(state: Boolean = true) {
    isLookLocked = state
  }

  @JvmStatic
  fun setMovementLock(state: Boolean = true) {
    isMovementLocked = state
    if (!state) {
      clearForcedMovement()
    }
  }

  @JvmStatic
  fun setForcedMovement(
    forward: Boolean,
    backward: Boolean,
    @Suppress("UNUSED_PARAMETER")
    left: Boolean,
    @Suppress("UNUSED_PARAMETER")
    right: Boolean,
    jump: Boolean,
    shift: Boolean,
    sprint: Boolean
  ) {
    movementOwner = MovementOwner.PATHFINDER
    isMovementLocked = true

    forcedForward = forward
    forcedBackward = backward
    forcedLeft = left
    forcedRight = right
    forcedJump = jump
    forcedShift = shift
    forcedSprint = sprint

    hasForcedMovement = true
  }

  @JvmStatic
  fun setForcedMovement(
    owner: MovementOwner,
    forward: Boolean,
    backward: Boolean,
    left: Boolean,
    right: Boolean,
    jump: Boolean,
    shift: Boolean,
    sprint: Boolean
  ): Boolean {
    if (!requestMovement(owner)) return false

    forcedForward = forward
    forcedBackward = backward
    forcedLeft = left
    forcedRight = right
    forcedJump = jump
    forcedShift = shift
    forcedSprint = sprint

    hasForcedMovement = true
    return true
  }

  @JvmStatic
  fun setForcedActions(
    owner: MovementOwner,
    attack: Boolean = false,
    use: Boolean = false
  ): Boolean {
    if (!requestMovement(owner)) return false

    forcedActionsEnabled = true
    forcedAttack = attack
    forcedUse = use

    return true
  }

  @JvmStatic
  fun clearForcedMovement() {
    ForcedInputBridge.releasePhysicalKeys()
    hasForcedMovement = false
    forcedForward = false
    forcedBackward = false
    forcedLeft = false
    forcedRight = false
    forcedJump = false
    forcedShift = false
    forcedSprint = false
    forcedAttack = false
    forcedUse = false
    forcedActionsEnabled = false
  }

  @JvmStatic
  fun clearForcedMovement(owner: MovementOwner) {
    if (owner != MovementOwner.PATHFINDER && owner != MovementOwner.FAILSAFE && movementOwner != owner) return
    clearForcedMovement()
  }

  @JvmStatic
  fun clearAll(owner: MovementOwner) {
    releaseMovement(owner)
    releaseLook(owner)
  }

  @JvmStatic
  fun forceClearAll() {
    movementOwner = MovementOwner.NONE
    lookOwner = MovementOwner.NONE
    isMovementLocked = false
    isLookLocked = false
    clearForcedMovement()
  }

  @JvmStatic
  fun currentMovementOwnerName(): String = movementOwner.name

  @JvmStatic
  fun currentLookOwnerName(): String = lookOwner.name

  @JvmStatic
  fun forceRecoveryJump(): Boolean = setForcedMovement(
    owner = MovementOwner.RECOVERY,
    forward = true, backward = false, left = false, right = false,
    jump = true, shift = false, sprint = false
  )

  @JvmStatic
  fun forceRecoveryBackup(): Boolean = setForcedMovement(
    owner = MovementOwner.RECOVERY,
    forward = false, backward = true, left = false, right = false,
    jump = false, shift = false, sprint = false
  )

  @JvmStatic
  fun forceRecoveryStrafeLeft(): Boolean = setForcedMovement(
    owner = MovementOwner.RECOVERY,
    forward = false, backward = false, left = true, right = false,
    jump = false, shift = false, sprint = false
  )

  @JvmStatic
  fun forceRecoveryStrafeRight(): Boolean = setForcedMovement(
    owner = MovementOwner.RECOVERY,
    forward = false, backward = false, left = false, right = true,
    jump = false, shift = false, sprint = false
  )
}
